### PR TITLE
fix Bug in BertScore calculation: pred target misalignment

### PR DESCRIPTION
**Pull request by gxy-gxy**
## fix Bug in BertScore calculation: pred target misalignment

Fixes bug in BertScore cal.
This pull request addresses a bug identified in the BertScore calculation within the TextDataset class in `src/torchmetrics/functional/text/helper_embedding_metric.py`.
The class is designed with a preprocess function automatically sorts input text by length to optimize batch encoding efficiency. However, this behavior introduces an issue during the BertScore calculation process, as predictions (preds) and targets (targets) are initialized in separate datasets. This results in a mismatched ordering of text pairs, which is problematic given the pairwise nature of BertScore's calculation. To ensure accurate scoring, it is critical to re-align the datasets to their original order before computing the scores. The proposed fix involves ensuring that the datasets for predictions and targets are processed in a way that maintains their original pairing throughout the calculation process.

Here is the fixed code:
``` 
preds_embeddings = preds_embeddings[preds_loader.dataset.sorting_indices]
target_embeddings = target_embeddings[target_loader.dataset.sorting_indices]
```

This change is essential for preserving the integrity of the BertScore evaluation, ensuring that each prediction is accurately compared against its corresponding target.

----
📚 Documentation preview 📚: https://torchmetrics--2347.org.readthedocs.build/en/2347/

### CODE REVIEW

**Comment by gxy-gxy** [comment]
Here is the test code:
```
from torchmetrics.text.bert import BERTScore

score_model = BERTScore(model_name_or_path='roberta-large', batch_size=2)

text1 = [ "Claim A from machine", "Claim A from machine"]
text2 = ["Claim A from machine", "Claim B"]
similarities = score_model(text1, text2)
print(similarities)
```

**Comment by Borda** [comment]
@stancld could you help here, pls?

**Review by baskrahmer — COMMENTED** [review]
Good catch! I can reproduce the bug using your test code snippet. Although the actual metric values seem to be correct, the ordering is not always valid.

It might be nice to somehow integrate this case with the current test suite. E.g. an assertation that reversing the targets/preds also reverses the scores.

File: src/torchmetrics/functional/text/bert.py [hunk 1]
```diff
@@ -419,18 +419,12 @@ def bert_score(
         preds_loader, preds_dataset.max_length, model, device, num_layers, all_layers, idf, verbose, user_forward_fn
     )
 
+    preds_embeddings = preds_embeddings[preds_loader.dataset.sorting_indices]
+    target_embeddings = target_embeddings[target_loader.dataset.sorting_indices]
```
[the remark this answers is not in the document]
Comment (gxy-gxy) [reply src/torchmetrics/functional/text/bert.py:423]: > I'm pretty sure the IDF factors also need to be unsorted too

Thanks for your suggestion, I will fix it soon.

[the remark this answers is not in the document]
Comment (Borda) [reply src/torchmetrics/functional/text/bert.py:423]: @gxy-gxy how id it going, will you have time to update it?

**Comment by Borda** [comment]
> Here is the test code:
> 
> ```python
> from torchmetrics.text.bert import BERTScore
> 
> score_model = BERTScore(model_name_or_path='roberta-large', batch_size=2)
> 
> text1 = [ "Claim A from machine", "Claim A from machine"]
> text2 = ["Claim A from machine", "Claim B"]
> similarities = score_model(text1, text2)
> print(similarities)
> ```

@gxy-gxy can you pls add it as a test?

**Review by Borda — CHANGES_REQUESTED** [review]
https://github.com/Lightning-AI/torchmetrics/pull/2347#issuecomment-2230588194

File: src/torchmetrics/functional/text/bert.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gxy-gxy) [reply src/torchmetrics/functional/text/bert.py:423]: hello， I have fixed it.

**Review by baskrahmer — APPROVED** [review]
LGTM! Thanks for the addition

**Comment by Borda** [comment]
@baskrahmer, would mind also adding an entry to the changelog?

**Pull request merged**